File: python/ten_minute_walk.py
# https://www.codewars.com/kata/54da539698b8a2ad76000228


def is_valid_walk(walk: list[str]):
    if len(walk) != 10:
        return False

    dirs = {"n": 0, "s": 0, "e": 0, "w": 0}
    for step in walk:
        if step not in walk:
            raise ArgumentError()
        dirs[step] += 1

    return dirs["n"] == dirs["s"] and dirs["e"] == dirs["w"]

File: haskell/ten_minute_walk.hs
-- https://www.codewars.com/kata/54da539698b8a2ad76000228

import           Testing (assert)

isValidWalk :: [Char] -> Bool
isValidWalk walk = length (take 11 walk) == 10 && (count 'n' walk == count 's' walk) && (count 'e' walk == count 'w' walk)
  where
    count :: Char -> [Char] -> Int
    count step = length . filter (==step)


tests :: IO()
tests = do
  assert True (isValidWalk ['n','s','n','s','n','s','n','s','n','s'])
  assert False (isValidWalk ['n','s','n','s','n','s','n','s','n','n'])
  assert False (isValidWalk ['n','s'])
  assert False (isValidWalk (repeat 'n'))
  assert True (isValidWalk ['n','s','e','w','n','s','e','w','n','s'])
